{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "936132bb",
   "metadata": {},
   "source": [
    "# Create Pangenome Fasta File\n",
    "\n",
    "This notebook walks through the steps required to create a fasta file from a set of haplotypes in a Practical Haplotype Graph. This file is required to impute haplotypes from sequencing data.\n",
    "\n",
    "This notebook assumes that you have a Practical Haplotype Graph database and have loaded haplotypes from your desired taxa (See steps 1 and 2 for constructing and populating the database).\n",
    "\n",
    "## How It Works\n",
    "\n",
    "WriteFastaFromGraphPlugin takes a haplotype graph as input and writes the sequences for each haplotype in the graph to a fasta file. The name of each sequence is an integer equal to its haplotype_id within the database. This \"pangenome\" fasta file will likely be several times larger than your reference genome fasta.\n",
    "\n",
    "In order to provide the input for WriteFastaFromGraphPlugin, another plugin called HaplotypeGraphBuilderPlugin must be run first. HaplotypeGraphBuilderPlugin constructs a graph from user-specified methods, taxa, and chromosomes. Methods include haplotypes generated directly from assemblies or WGS data as well as consensus haplotypes generated by RunHapConsensusPipelinePlugin (See notebook for details). \n",
    "\n",
    "## Requirements\n",
    "\n",
    "- A Practical Haplotype Graph database, with taxa loaded and (optionally) consensus haplotypes generated.\n",
    "- A config file with database connection information and, optionally, Tassel plugin parameters set.\n",
    "\n",
    "### Required Parameters\n",
    "\n",
    "#### HaplotypeGraphBuilderPlugin\n",
    "\n",
    " - configFile: the config file containing database connection information\n",
    " - methods: Pairs of methods (haplotype method name and range group method name). Method pair separated by a comma, and pairs separated by colon. The range group is optional. Usage: <haplotype method name1>,<range group name1>:<haplotype method name2>,<range group name2>:<haplotype method name3>\n",
    "\n",
    "#### WriteFastaFromGraphPlugin\n",
    "\n",
    "- outputFile: path to the output pangenome fasta file\n",
    "\n",
    "### Optional Parameters (default in parentheses)\n",
    "\n",
    "#### HaplotypeGraphBuilderPlugin\n",
    "\n",
    "- includeSequences (true): whether to include sequences in haplotype nodes. For creating the pangenome, this should be left as the default true\n",
    "- includeVariantContexts (false): whether to include variant contexts in haplotype nodes. For creating the pangenome, this should be left as the default false\n",
    "- haplotypeIDs: A list of haplotype ids to include in the graph. If not specified, all IDs are included\n",
    "- chromosomes: A list of chromosomes to include in the graph. If not specified, all chromosomes are included\n",
    "- taxa: A list of taxa to include in the graph. This can be a comma separated list of taxa (no spaces unless surrounded by quotes), file (.txt) with list of taxa names to include, or a taxa list file (.json or .json.gz). By default, all taxa will be included.\n",
    "- localGVCFFolder: folder where the reference and assembly gvcfs are stored. Only required if includeVariantContexts is true."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "107b28ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########\n",
    "# EDIT ME #\n",
    "###########\n",
    "\n",
    "# working directory\n",
    "working_dir = \"/workdir/ahb232/phg_sorghum_apr2023/\"\n",
    "\n",
    "# location of the config file, relative to working_dir\n",
    "config_file = \"config.txt\"\n",
    "\n",
    "# path to the output pangenome fasta file, relative to working_dir\n",
    "pangenome_file = \"/outputDir/pangenome/pangenome.fa\"\n",
    "\n",
    "# location of the log file. To write the log to this notebook, use log_file = \"\"\n",
    "log_file = \"/logs/create_pangenome_log.txt\"\n",
    "\n",
    "# list of methods to use to create the pangenome\n",
    "METHODS = \"HudsonAlpha_assembly:public_assembly\"\n",
    "\n",
    "# docker command. Usually \"docker\", but on bioHPC should be \"docker1\"\n",
    "DOCKER = \"docker1\"\n",
    "\n",
    "# docker version\n",
    "DOCKER_VERSION = \"maizegenetics/phg:1.4\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8e66d21",
   "metadata": {},
   "outputs": [],
   "source": [
    "## RUN THIS CODE BLOCK BUT DO NOT EDIT ##\n",
    "\n",
    "CONFIG = \"/phg/\" + config_file\n",
    "\n",
    "FASTA_FILE = \"/phg/\" + pangenome_file\n",
    "\n",
    "TO_LOG = \"\"\n",
    "\n",
    "if (log_file != \"\"):\n",
    "    TO_LOG = \" > \" + working_dir + \"/\" + log_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a5a13ff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "! {DOCKER} run --name create_pangenome --rm \\\n",
    "    -v {working_dir}/:/phg/ \\\n",
    "    -t {DOCKER_VERSION} \\\n",
    "    /tassel-5-standalone/run_pipeline.pl -Xmx100G -debug -configParameters {CONFIG} \\\n",
    "    -HaplotypeGraphBuilderPlugin \\\n",
    "    -configFile {CONFIG} \\\n",
    "    -methods {METHODS} \\\n",
    "    -endPlugin \\\n",
    "    -WriteFastaFromGraphPlugin \\\n",
    "    -outputFile {FASTA_FILE} \\\n",
    "    -endPlugin {TO_LOG}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3fdca86",
   "metadata": {},
   "source": [
    "## Indexing the Pangenome\n",
    "\n",
    "The next plugin in the imputation pipeline, FastqToMappingPlugin expects a minimap2 index (.mmi) file. Run the code blocks below to generate it. A .mmi file may not be necessary if you want to align your reads outside of the PHG, so this step is optional."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "21078706",
   "metadata": {},
   "outputs": [],
   "source": [
    "# EDIT ME #\n",
    "\n",
    "# path to minimap index file, relative to working_dir\n",
    "index_path = \"/outputDir/pangenome/pangenome.mmi\"\n",
    "\n",
    "# Minimap2 index parameter k, the kmer length of the minimizers, \n",
    "# which is used to index the pangenome.\n",
    "KMER_LENGTH = 21\n",
    "\n",
    "# Minimap2 index parameter I, the maximum number of bases loaded into memory, \n",
    "# which is used to index the pangenome.\n",
    "# This must be large enough to hold the entire pangenome in memory.\n",
    "NUM_BASES = \"90G\"\n",
    "\n",
    "# Minimap2 index parameter w, the minimizer window size, which is used to index the pangenome.\n",
    "WINDOW_SIZE = 11\n",
    "\n",
    "INDEX_PATH = \"/phg/\" + index_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d78b5a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "! {DOCKER} run --name index_pangenome --rm \\\n",
    "    -v {working_dir}/:/phg/ \\\n",
    "    -t {DOCKER_VERSION} \\\n",
    "    minimap2 -d {INDEX_NAME} -k {KMER_LENGTH} -I {NUM_BASES} -w {WINDOW_SIZE} {FASTA_FILE}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1673ee43",
   "metadata": {},
   "source": [
    "## Next Steps\n",
    "\n",
    "Want to continue with the Imputation Pipeline? The next step is to align your low-coverage sequence data to the fasta or .mmi file created with this notebook and map reads to haplotypes. The PHG has a plugin to do so, called FastqToMappingPlugin (see FastaToMapping), which uses minimap2 to align reads. Alignment can also be done outside the PHG, producing sam files which are converted to read mappings by SAMToReadMappingPlugin. This may be useful in situations where you would like to perform the resource-intensive alignment step on a machine that cannot easily connect to your PHG database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85f766e4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
