{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eff48b0d",
   "metadata": {},
   "source": [
    "# FASTQ To Read Mapping\n",
    "\n",
    "This notebook describes the procedure required to run FastqToMappingPlugin with a Practical Haplotype Graph. It uses minimap2 to align reads to the pangenome, then assigns reads to haplotypes in the reference range with the most hits. \n",
    "\n",
    "This notebook assumes you are running the plugin within Docker using the image maizegenetics/phg version 1.4 or later.\n",
    "\n",
    "## Requirements\n",
    "\n",
    "- A Practical Haplotype Graph database with the desired haplotypes loaded (see steps 1 and 2 for setting up the database)\n",
    "- Haplotype method(s) to map reads to. This can be either a set of haplotypes loaded directly from assemblies or WGS data, or a set of consensus haplotypes (see CreateConsensiNotebook for details)\n",
    "- A config file containing database connection information and (optionally) plugin parameters\n",
    "- A keyfile. See below for details.\n",
    "\n",
    "## Keyfile\n",
    "\n",
    "The keyfile for this plugin describes the sets of fastq reads to be mapped. It is a tab-separated table with a required header with the following columns:\n",
    "- cultivar: (required) Name of the taxon to be processed\n",
    "- flowcell_lane: (required) Name of the flow cell this sample came from\n",
    "- filename: (required) Name of the fastq file to be processed. Do not include paths\n",
    "- filename2: (required if using paired end reads) Second fastq file to be processed\n",
    "- PlateID: (optional) The ID of the plate\n",
    "\n",
    "Cultivar + flowcell_lane + PlateID should be unique for each sample in the keyfile. \n",
    "\n",
    "## Parameters\n",
    "\n",
    "### HaplotypeGraphBuilderPlugin Parameters\n",
    "\n",
    "### Required\n",
    "\n",
    "- configFile: Database configuration file\n",
    "- methods: Pairs of methods (haplotype method name and range group method name) Method pair separated by a comma, and pairs separate by colon. The range group is optional. Usage: <haplotype method name 1>,<range group name 1>:<haplotype method name 2>,<range group name 2>:<haplotype method name 3>\n",
    "\n",
    "### Optional (default in parentheses)\n",
    "\n",
    "- includeSequences: (true) Whether to include sequences in haplotype nodes\n",
    "- includeVariantContexts: (false) Whether to include variant contexts in haplotype nodes. Variant contexts are not required for FastqToMappingPlugin.\n",
    "- haplotypIds: List of haplotype IDs to include in the graph. If not specified, all IDs are included.\n",
    "- chromosomes: List of chromosomes to include in graph. Defailt is to include all chromosomes.\n",
    "- taxa: List of taxa to include. Default is to include all taxa.\n",
    "- localGVCFFolder: Folder where reference/assembly GVCFs are stored. Only required if includeVariantContexts is true.\n",
    "\n",
    "\n",
    "### FastqToMappingPlugin Parameters\n",
    "\n",
    "\n",
    "### Required\n",
    "\n",
    "- minimap2IndexFile: minimap2 index file foe the pangenome\n",
    "- keyFile: keyFile with the list of files to process. See above for formatting.\n",
    "- fastqDir: directory containing the fastq files\n",
    "- methodName: a unique name for this read mapping method, to be stored in the database\n",
    "\n",
    "### Optional (default in parentheses)\n",
    "\n",
    "- maxRefRangeErr: (.25) Maximum allowed error when choosing the best reference range to count. Error is defined as 1-(mostHitRefRangeCount / totalHitCount)\n",
    "- lowMemMode: (true) Run in low memory mode.\n",
    "- maxSecondary: (20) Maximum number of secondary alignments to be returned by minimap2. This will be the value of the -N parameter in minimap2 command line.\n",
    "- fParameter: (f1000,5000) The f parameters used by minimap2. If the sr preset (-x sr) is used then thi parameters takes the form f<int1,int2>. From the minimap2 map page: \"If integer, ignore minimizers occuring more than INT1 times. INT2 is only effective in the --sr or -xsr mode, which sets the threshold for a second round of seeding.\"\n",
    "- minimapLocation: (minimap2) location of minimap2 executable\n",
    "- methodDescription: Method description to be stored in the database\n",
    "- debugDir: Directory to write out the read mapping files. This is optional for debug purposes.\n",
    "- outputSecondaryStats: (false) Output secondary mapping statistics such as total AS for each haplotype ID\n",
    "- isTestMethod: (false) Indication if the data is to be loaded against a test method Data loaded with test methods are not cached with the PHG ktor server.\n",
    "- updateDB: (true) If se to true, the read mappings will be written to the db. Otherwise nothing will be written.\n",
    "- runWithoutGraph: (false) If true, will require the input of a JSON file created by CreateHapIdMapsPlugin and will not require an input Graph object from HaplotypeGraphBuilderPlugin. Useful when running alignments on a machine that cannot connect to the database.\n",
    "- hapIDMapFile: Location of the HapIdMapFile where the graph information can be found. Required in runWithoutGraph is true\n",
    "- inputFastqFile: Input of the first fastq file to be run. If the keyfile has a pair it will be detected. If this is not set, everything in the input directory will be run. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3efa2f38",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########\n",
    "# EDIT ME #\n",
    "###########\n",
    "\n",
    "# Path to the working directory\n",
    "working_dir = \"/workdir/ahb232/phg_sorghum_apr2023/\"\n",
    "\n",
    "# config file path (relative to working_dir)\n",
    "config = \"config.txt\"\n",
    "\n",
    "# path to key file (relative to working_dir)\n",
    "key_file = \"IS36143_keyfile.txt\"\n",
    "\n",
    "# methods to pass to HaplotypeGraphBuilderPlugin (see above for details)\n",
    "haplotype_methods = \"HudsonAlpha_assembly:public_assembly\"\n",
    "\n",
    "# location of the pangenome minimap index file (relative to working_dir)\n",
    "minimap_index_file = \"/outputDir/pangenome/pangenome.mmi\"\n",
    "\n",
    "# folder (relative to working_dir) containing the fastq reads to map to the pangenome \n",
    "fastq_dir = \"/WGS/sorghumbase/downsampled/\"\n",
    "\n",
    "# name and (optional) description for the mapping method\n",
    "mapping_method = \"IS36143_downsampled_WGS_readmapping\"\n",
    "mapping_method_description = \"\"\n",
    "\n",
    "# path to the log file (relative to working_dir)\n",
    "# Note: You may set log_file to \"\", in which case the log will print to the notebook cell. \n",
    "# This may make the notebook slow to open if the log is very long.\n",
    "log_file = \"/logs/readmapping_IS36143_log.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2c54706",
   "metadata": {},
   "outputs": [],
   "source": [
    "# RUN BUT DO NOT EDIT #\n",
    "\n",
    "CONFIG = \"/phg/\" + config\n",
    "KEYFILE = \"/phg/\" + key_file\n",
    "MMI_FILE = \"/phg/\" + minimap_index_file\n",
    "FASTQ_DIR = \"/phg/\" + fastq_dir\n",
    "\n",
    "TO_LOG = \"\"\n",
    "\n",
    "if (log_file != \"\"):\n",
    "    TO_LOG = \" > \" + LOG_FILE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1eee8a5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "! {DOCKER} run --name create_pangenome --rm \\\n",
    "    -v {working_dir}/:/phg/ \\\n",
    "    -t {DOCKER_VERSION} \\\n",
    "    /tassel-5-standalone/run_pipeline.pl -Xmx100G -debug -configParameters {CONFIG} \\\n",
    "    -HaplotypeGraphBuilderPlugin \\\n",
    "    -configFile {CONFIG} \\\n",
    "    -methods {haplotype_methods} \\\n",
    "    -endPlugin \\\n",
    "    -FastqToMappingPlugin \\\n",
    "    -minimap2IndexFile {MMI_FILE} \\\n",
    "    -keyFile {KEYFILE} \\\n",
    "    -fastqDir {FASTQ_DIR} \\\n",
    "    -methodName {mapping_method} \\\n",
    "    -methodDescription {mapping_method_description} \\\n",
    "    -endPlugin {TO_LOG}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
