{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "93607670",
   "metadata": {},
   "source": [
    "# GVCF Alignment Quality\n",
    "\n",
    "This notebook calculates and plots features relevant to alignment quality - such as number of SNPs, number of indels, and alignment coverage - for a set of gvcf files. These gvcf files are assumed to have been produced by aligning whole assemblies and/or WGS data to a reference genome.\n",
    "\n",
    "## Requirements\n",
    "\n",
    "* A directory containing one or more gvcf files, which may be gzipped\n",
    "* Docker\n",
    "* Python packages: matplotlib, numpy, pandas, itables, seaborn\n",
    "\n",
    "## Configuration and User Variables\n",
    "\n",
    "Run the following cells to set up your environment. In the cell labeled \"Edit Me\", change the directories to match your filesystem. Notice that Docker-specific filepaths may differ from the absolute filepath on your machine.\n",
    "\n",
    "Later in this notebook, you will see other cells labeled \"Edit Me\". Take a moment to look over the variables in these cells and change them to suit your analysis. \n",
    "\n",
    "It is recommended that you run all cells in order. Some cells depend on the results from previous cells, and they will throw errors if you skip them. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b15f4419",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import required libraries\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas\n",
    "from collections import Counter\n",
    "import math\n",
    "from itables import init_notebook_mode\n",
    "import seaborn as sns\n",
    "\n",
    "tmp = init_notebook_mode(all_interactive=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3eedfa63",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########\n",
    "# EDIT ME #\n",
    "###########\n",
    "\n",
    "# working directory\n",
    "working_dir = \"/path/to/working/directory/\"\n",
    "\n",
    "# Note: for the following variables, please use the path relative to the working directory\n",
    "# The reason for this is that the absolute path changes within the docker\n",
    "\n",
    "# the directory containing the gvcf files\n",
    "gvcf_dir = \"/gvcf/\"\n",
    "\n",
    "# the next two output files can also be nested within a subdirectory\n",
    "# make sure that directory exists before continuing\n",
    "\n",
    "# the main output file - a table of summary statistics\n",
    "summary_file_name = \"metrics/sorghum_gvcf_metrics.tsv\"\n",
    "\n",
    "# optional output file - lists the sizes of indels\n",
    "# if you don't want this output, use indels_file_name = \"\"\n",
    "indels_file_name = \"metrics/sorghum_gvcf_indels.tsv\"\n",
    "\n",
    "# path to figure directory, relative to working_dir\n",
    "fig_dir = \"/figures/\"\n",
    "\n",
    "# Docker-specific parameters\n",
    "\n",
    "# Docker image and version\n",
    "DOCKER_VERSION=\"maizegenetics/phg:latest\"\n",
    "# Docker path - location where Docker is installed\n",
    "# Note: on CBSU machines this is docker1\n",
    "DOCKER_PATH=\"docker\"\n",
    "\n",
    "# Do not change these; they are docker-specific paths\n",
    "PHG_VCF_DIR=\"/phg/\" + gvcf_dir\n",
    "PHG_OUT_FILE=\"/phg/\" + summary_file_name\n",
    "\n",
    "if indels_file_name == \"\":\n",
    "    PHG_INDEL_FLAG = \"\"\n",
    "else:\n",
    "    PHG_INDEL_FLAG = \"-indelFile /phg/\" + indels_file_name "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e0b556f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make figure folder if it doesn't exist\n",
    "\n",
    "if not os.path.exists(working_dir + \"/\" + fig_dir):\n",
    "    os.mkdir(working_dir + \"/\" + fig_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73e87aa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run VCFMetricsPlugin\n",
    "\n",
    "! {DOCKER_PATH} run --name vcf_metrics --rm \\\n",
    "    -v {working_dir}/:/phg/ \\\n",
    "    -t {DOCKER_VERSION} \\\n",
    "    /tassel-5-standalone/run_pipeline.pl -Xmx50G -debug \\\n",
    "    -VCFMetricsPlugin \\\n",
    "        -vcfDir {PHG_VCF_DIR} \\\n",
    "        -outFile {PHG_OUT_FILE} \\\n",
    "        {PHG_INDEL_FLAG} \\\n",
    "    -endPlugin"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "917b7098",
   "metadata": {},
   "source": [
    "# Summary Table\n",
    "\n",
    "The cell below displays the summary table that the plugin writes. Summaries for each assembly are on rows that begin with \\[assembly name\\]\\_ALL. Following the overall summary are rows containing overviews of each chromosome represented in the gvcf. This notebook shows only the whole-assembly summary rows. \n",
    "\n",
    "You may wish to add more columns to this table. For example, a \"shortNames\" column could contain abbreviated assembly names for cleaner labeling on figures. Or, a \"group\" column might contain metadata about the technology used to produce the assemblies, so that points on a graph may be grouped and/or color-coded. The table can be opened and edited in Excel or any simple text editing program. If using a basic text editor, make sure to separate columns by tabs and not spaces. If you add columns to the table, please re-run the cell below so that the new columns appear."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3132ee6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load in metrics table, and save summary lines to summary_table\n",
    "metrics_table = pandas.read_csv(working_dir + \"/\" + summary_file_name, sep=\"\\t\", index_col=False)\n",
    "\n",
    "summary_table = metrics_table[[name.endswith(\"_ALL\") for name in metrics_table[\"name\"]]]\n",
    "summary_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa80c513",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########\n",
    "# EDIT ME #\n",
    "###########\n",
    "\n",
    "# Display these assembly names on figures. Default is sample_names = summary_table[\"name\"]\n",
    "\n",
    "# To change use a column in the summary table (you may directly edit the .tsv file to add columns)\n",
    "# Select the appropriate column with the format: sample_names = summary_table[\"column_name\"]\n",
    "sample_names = summary_table[\"name\"]\n",
    "\n",
    "# If you wish to color-code data points based on a categorical variable you supply (e.g. alignment software), \n",
    "# specify the groups here. Like sample_names, add a column to the .tsv file\n",
    "# If you do not wish to color-code, set sample_groups = list()\n",
    "\n",
    "sample_groups = list()\n",
    "\n",
    "\n",
    "# Do not edit this. Having the sample names/groups as lists makes some things easier later on\n",
    "sample_names = list(sample_names)\n",
    "sample_groups = list(sample_groups)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10e0f82e",
   "metadata": {},
   "source": [
    "# Percent Alignment and Mapping\n",
    "\n",
    "The following cell shows the percentage of bases aligned/mapped relative to the reference for each gvcf in the selected directory. In this context, \"aligned\" and \"mapped\" have separated definitions. A base is considered to be aligned if the base in the reference corresponds to a single base in the query. The bases do not need to be identical - SNPs count as well as reference blocks. In the exaple alignment below, there are 10 aligned bases - 7 that are the same as the reference, and 3 SNPs.\n",
    "\n",
    "``ref: GATC--GGATCAAAC``  \n",
    "``     | ||  ||   |  |``  \n",
    "``qry: GGTCTCGG---ATGC``\n",
    "\n",
    "Every base that is aligned is also mapped under this notebook's definition, but mapped bases also include insertions and deletions of any size. The only bases that are considered \"unmapped\" are those that are not included in any gvcf record, whether as part of a reference block or a variant. We have usually seen unmapped bases around the beginnings and ends of contigs. \n",
    "\n",
    "All percentages below are relative to the reference genome, i.e. the percentage of bases in the reference genome that are aligned or mapped."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64084541",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot percent of base pairs aligned/mapped\n",
    "\n",
    "percents = summary_table[\"percentRefAligned\"].tolist()\n",
    "percents.extend(summary_table[\"percentRefMapped\"].tolist())\n",
    "\n",
    "data = pandas.DataFrame.from_dict({\"name\": sample_names + sample_names,\n",
    "        \"percent\": percents, \n",
    "        \"count\": [\"aligned\"] * len(sample_names) + [\"mapped\"] * len(sample_names)})\n",
    "\n",
    "if not len(sample_groups) == 0:\n",
    "    data[\"group\"] = sample_groups + sample_groups\n",
    "    data.sort_values(by=[\"group\", \"percent\"], inplace=True, ascending=[True, True])\n",
    "else:\n",
    "    data.sort_values(by=\"percent\", inplace=True, ascending=False)\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "if len(sample_groups) == 0:\n",
    "    sns.scatterplot(data=data, x=\"name\", y=\"percent\", ax=ax, style=\"count\", markers=[\"o\", \"s\"])\n",
    "else:\n",
    "    sns.scatterplot(data=data, x=\"name\", y=\"percent\", hue=\"group\", ax=ax, style=\"count\", markers=[\"o\", \"s\"])\n",
    "\n",
    "#Format axis\n",
    "ax.set_ylim([0, 1])\n",
    "ax.set_yticks([0, 0.2, 0.4, 0.6, 0.8, 1])\n",
    "ax.set_yticklabels([\"0%\", \"20%\", \"40%\", \"60%\", \"80%\", \"100%\"])\n",
    "ax.set_xticks(ax.get_xticks())\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=-90, ha=\"right\")\n",
    "ax.set(xlabel=None, ylabel=None)\n",
    "ax.set_axisbelow(True)\n",
    "\n",
    "plt.grid()\n",
    "\n",
    "plt.title(\"Percentage of bases aligned and mapped to the reference genome\")\n",
    "\n",
    "# Can change this parameter if bottom test gets cut off\n",
    "plt.gcf().subplots_adjust(bottom=0.25)\n",
    "plt.legend(bbox_to_anchor=(1.0, 1.0))\n",
    "\n",
    "fig.savefig(working_dir + \"/\" + fig_dir + \"/\" + \"percent_bases_aligned.svg\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f3fa2f5",
   "metadata": {},
   "source": [
    "# Number of SNPs\n",
    "\n",
    "The following cell generates a figure showing the number of SNPs present in each GVCF file. SNPs may include real SNPs as well as sequencing errors and missing bases. Indels and unmapped regions are not included."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25e8a964",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot number of SNPs\n",
    "\n",
    "data = pandas.DataFrame.from_dict({\n",
    "    \"name\": sample_names,\n",
    "    \"snps\": summary_table[\"numSNPs\"]\n",
    "})\n",
    "\n",
    "if len(sample_groups) == 0:\n",
    "    data.sort_values(by=[\"snps\"], inplace=True, ascending=[True])\n",
    "else:\n",
    "    data[\"group\"] = sample_groups\n",
    "    data.sort_values(by=[\"group\", \"snps\"], inplace=True, ascending=[True, True])\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "if len(sample_groups) == 0:\n",
    "    sns.barplot(data=data, x=\"name\", y=\"snps\", dodge=False)\n",
    "else:\n",
    "    sns.barplot(data=data, x=\"name\", y=\"snps\", hue=\"group\", dodge=False)\n",
    "\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=-90, ha=\"right\")\n",
    "\n",
    "ax.set(xlabel=\"count\", ylabel=None)\n",
    "ax.set_axisbelow(True)\n",
    "ax.get_yaxis().set_major_formatter(matplotlib.ticker.FuncFormatter(lambda x, p: format(int(x), ',')))\n",
    "\n",
    "plt.grid()\n",
    "\n",
    "plt.title(\"Number of SNPs relative to reference\")\n",
    "plt.gcf().subplots_adjust(bottom=0.25, left=0.15)\n",
    "plt.legend(bbox_to_anchor=(1.0, 1.0))\n",
    "fig.savefig(working_dir + \"/\" + fig_dir + \"/\" + \"num_snps.svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca7fc072",
   "metadata": {},
   "source": [
    "# Number of Indels\n",
    "\n",
    "The following cell generates a figure showing the number of indels in each gvcf. Indels are defined here as variants where the length of the reference and the length of the alternate are different. They do not include unmapped regions. \n",
    "In this graph, the number of insertions and deletions are separated and shown as a stacked bar (deletions in lighter color on top, insertions in darker color on bottom)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c68f23aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pandas.DataFrame({\n",
    "    \"total\": summary_table[\"numIndels\"].tolist()\n",
    "}, index=sample_names)\n",
    "\n",
    "insertions = summary_table[\"numIns\"].tolist()\n",
    "deletions = summary_table[\"numDel\"].tolist()\n",
    "\n",
    "if len(sample_groups) == 0:\n",
    "    data[\"Ins\"] = insertions\n",
    "    data[\"Del\"] = deletions\n",
    "    \n",
    "    colors = iter([plt.cm.tab20(i) for i in range(2)])\n",
    "    cmap = [next(colors) for x in range(2)]\n",
    "    \n",
    "    data.sort_values(by=[\"total\"], inplace=True, ascending=[True])\n",
    "    data.drop([\"total\"], inplace=True, axis=1)\n",
    "else:    \n",
    "    \n",
    "    data[\"group\"] = sample_groups\n",
    "    tmp_counter = 0\n",
    "    for group in np.unique(sample_groups):\n",
    "        tmp_counter += 1\n",
    "        subgroup = group + \" - Ins\"\n",
    "        data[subgroup] = [insertions[x] if sample_groups[x] == group else 0 for x in range(len(sample_names))]\n",
    "        subgroup = group + \" - Del\"\n",
    "        data[subgroup] = [deletions[x] if sample_groups[x] == group else 0 for x in range(len(sample_names))]\n",
    "\n",
    "    \n",
    "    colors = iter([plt.cm.tab20(i) for i in range(tmp_counter * 2)])\n",
    "    cmap = [next(colors) for x in range(tmp_counter * 2)]\n",
    "\n",
    "    data.sort_values(by=[\"group\", \"total\"], inplace=True, ascending=[True, True])\n",
    "    data.drop([\"group\", \"total\"], inplace=True, axis=1)\n",
    "\n",
    "    \n",
    "data.plot(kind=\"bar\", stacked=True, color=cmap, zorder=2.5)\n",
    "\n",
    "ax = plt.gca()\n",
    "\n",
    "\n",
    "ax.set(xlabel=\"count\", ylabel=None)\n",
    "\n",
    "ax.get_yaxis().set_major_formatter(matplotlib.ticker.FuncFormatter(lambda x, p: format(int(x), ',')))\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=-90, ha=\"right\")\n",
    "\n",
    "plt.grid()\n",
    "\n",
    "plt.title(\"Number of indels relative to reference\")\n",
    "plt.legend(bbox_to_anchor=(1.0, 1.0))\n",
    "\n",
    "plt.gcf().subplots_adjust(bottom=0.25, left=0.15, right=1.0)\n",
    "\n",
    "plt.gcf().savefig(working_dir + \"/\" + fig_dir + \"/\" + \"num_indels.svg\")\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ea92ab4",
   "metadata": {},
   "source": [
    "# Largest Indels\n",
    "\n",
    "The next cell generates a figure showing the largest indel in each gvcf file. The shape of the point indicates whether that indel is an insertion or a deletion. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85921da4",
   "metadata": {},
   "outputs": [],
   "source": [
    "ins = summary_table[\"largestInsSize\"].tolist()\n",
    "delet = summary_table[\"largestDelSize\"].tolist()\n",
    "\n",
    "largest = summary_table[\"largestIndel\"].tolist()\n",
    "\n",
    "indels = [\"Insertion\" if ins[x] > delet[x] else \"Deletion\" for x in range(len(sample_names))]\n",
    "\n",
    "data = pandas.DataFrame.from_dict({\n",
    "    \"name\": sample_names,\n",
    "    \"type\": indels,\n",
    "    \"sizes\": largest\n",
    "})\n",
    "\n",
    "if len(sample_groups) == 0:\n",
    "    data.sort_values(by=[\"sizes\"], inplace=True, ascending=[True])\n",
    "else:\n",
    "    data[\"group\"] = sample_groups\n",
    "    data.sort_values(by=[\"group\", \"sizes\"], inplace=True, ascending=[True, True])\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "if len(sample_groups) == 0:\n",
    "    sns.scatterplot(data=data, x=\"name\", y=\"sizes\", ax=ax, style=\"type\", markers=[\"X\", \"o\"])\n",
    "else:\n",
    "    sns.scatterplot(data=data, x=\"name\", y=\"sizes\", hue=\"group\", ax=ax, style=\"type\", markers=[\"X\", \"o\"])\n",
    "\n",
    "ax.get_yaxis().set_major_formatter(matplotlib.ticker.FuncFormatter(lambda x, p: format(int(x), ',')))\n",
    "ax.set_xticks(ax.get_xticks())\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=-90, ha=\"right\")\n",
    "ax.set(xlabel=None, ylabel=\"size\")\n",
    "ax.set_axisbelow(True)\n",
    "\n",
    "plt.grid()\n",
    "plt.title(\"Largest indel sizes\")\n",
    "plt.gcf().subplots_adjust(bottom=0.25, left=0.15)\n",
    "plt.legend(bbox_to_anchor=(1.0, 1.0))\n",
    "\n",
    "fig.savefig(working_dir + \"/\" + fig_dir + \"/\" + \"largest_indels.svg\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
