{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c65b1c8a",
   "metadata": {},
   "source": [
    "# SAM Alignment Quality\n",
    "\n",
    "This notebook calculates and plots features relevant to alignment quality - such as number of mapped/unmapped reads and edit distance (smallest edit distance, if secondary alignments are present) - for a set of SAM files. \n",
    "\n",
    "## Requirements\n",
    "* A directory containing one or more .sam or .bam files. Records must contain the optional field \"NM\".\n",
    "* Docker\n",
    "* Python packages: matplotlib, ipyplot, pandas, scipy, numpy, itables\n",
    "\n",
    "## Configuration and User Variables\n",
    "\n",
    "Run the following three cells to set up your environment. In the cell labeled \"Edit Me\", change the directories to match your filesystem. Notice that Docker-specific filepaths are different from the absolute filepath on your machine."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c5bf1842",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.ticker as ticker\n",
    "import ipyplot\n",
    "import os\n",
    "import pandas\n",
    "import scipy.stats as st\n",
    "from scipy.stats import poisson\n",
    "from scipy.optimize import curve_fit\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "import math\n",
    "import random\n",
    "from itables import init_notebook_mode\n",
    "tmp = init_notebook_mode(all_interactive=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a8f3320",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########\n",
    "# EDIT ME #\n",
    "###########\n",
    "\n",
    "# Directories\n",
    "\n",
    "# Main directory\n",
    "working_dir = \"/path/to/working/directory/\"\n",
    "\n",
    "# Note: for the following directories, please use the path relative to the working directory\n",
    "# The reason for this is that the absolute path changes within the docker\n",
    "\n",
    "# SAM subfolder\n",
    "sam_dir=\"/sam/\"\n",
    "# Metrics subfolder - \n",
    "out_dir=\"/metrics/\"\n",
    "# Figures subfolder\n",
    "fig_dir=\"/figs/\"\n",
    "\n",
    "# Docker-specific parameters\n",
    "\n",
    "# Docker image and version\n",
    "DOCKER_VERSION=\"maizegenetics/phg:latest\"\n",
    "# Docker path - location where Docker is installed\n",
    "# Note: on CBSU machines this is docker1\n",
    "DOCKER_PATH=\"docker\"\n",
    "\n",
    "# Docker-specific directories - Do not edit directly\n",
    "PHG_SAM_DIR=\"/phg/\" + sam_dir\n",
    "PHG_OUT_DIR=\"/phg/\" + out_dir\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de2de1af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make metrics and figure directories, if they don't already exist\n",
    "\n",
    "if not os.path.exists(working_dir + \"/\" + out_dir):\n",
    "    os.mkdir(working_dir + \"/\" + out_dir)\n",
    "    \n",
    "if not os.path.exists(working_dir + \"/\" + fig_dir):\n",
    "    os.mkdir(working_dir + \"/\" + fig_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89ffdcaa",
   "metadata": {},
   "source": [
    "# Run Plugin\n",
    "\n",
    "First, we run the PHG plugin SAMMetricsPlugin to extract information from the SAM files. This step may take a few minutes per file, depending on their size. \n",
    "Alternatively, you may run the plugin outside this notebook: just make sure that the `-outDir` parameter points to the same directory as `out_dir` in the \"Edit Me\" cell above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "218748dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run SAMMetricsPlugin\n",
    "\n",
    "! {DOCKER_PATH} run --name sam_metrics --rm \\\n",
    "    -v {working_dir}/:/phg/ \\\n",
    "    -t {DOCKER_VERSION} \\\n",
    "    /tassel-5-standalone/run_pipeline.pl -Xmx50G -debug \\\n",
    "    -SAMMetricsPlugin \\\n",
    "        -samDir {PHG_SAM_DIR} \\\n",
    "        -outDir {PHG_OUT_DIR} \\\n",
    "    -endPlugin"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e923deb",
   "metadata": {},
   "source": [
    "# Summary Table and Plots\n",
    "\n",
    "Run the cells below to view a summary of the number of mapped and unmapped reads in each SAM file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e2fe81f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load table from file and print\n",
    "\n",
    "summary_table = pandas.read_csv(working_dir + \"/\" + out_dir + \"/SAM_summary_statistics.txt\", sep=\"\\t\")\n",
    "samples = [str(x) for x in summary_table[\"fileName\"]]\n",
    "\n",
    "# Calculate proportion of mapped reads\n",
    "\n",
    "summary_table[\"percentReadsMapped\"] = summary_table[\"mappedReads\"]/(summary_table[\"mappedReads\"] + summary_table[\"unmappedReads\"])\n",
    "summary_table = summary_table[[\"fileName\", \"mappedReads\", \"unmappedReads\", \"percentReadsMapped\", \"meanMinEditDistance\", \"medianMinEditDistance\"]]\n",
    "\n",
    "summary_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f45e822",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Generate a bar plot of the percentage of reads mapped for each SAM file\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.bar(samples, summary_table[\"percentReadsMapped\"])\n",
    "\n",
    "ax.set_ylim([0, 1])\n",
    "\n",
    "ax.set_yticks([0, 0.2, 0.4, 0.6, 0.8, 1])\n",
    "ax.set_yticklabels([\"0%\", \"20%\", \"40%\", \"60%\", \"80%\", \"100%\"])\n",
    "ax.set_xticks(ax.get_xticks())\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha=\"right\")\n",
    "ax.set_axisbelow(True)\n",
    "plt.grid()\n",
    "\n",
    "plt.title(\"Percentage of Reads Mapped\")\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9654ade2",
   "metadata": {},
   "source": [
    "# Distribution of Alignment Edit Distances\n",
    "\n",
    "For each fasta/fastq read that is successfully aligned to the reference sequence, minimap2 gives an edit distance - the number of single base insertions, deletions, or substitutions required to transform the query sequence into the reference sequence. An edit distance of 0 means that the query sequence perfectly matched the reference sequence. \n",
    "The following cells generate histograms showing the distribution of the edit distances for each mapped read in the SAM files (choosing the smallest edit distance if multiple alignments were generated for each read). The plots are saved to the `fig_dir` directory for future display/reference. In the \"Edit Me\" cell below, set parameters to change the look of the plots. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e767d997",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########\n",
    "# EDIT ME #\n",
    "###########\n",
    "\n",
    "# figure prefix - this is appended to the start of the figure file name when it is saved\n",
    "# use this if you want to generate multiple versions of the figures and keep them in the same folder\n",
    "fig_prefix = \"\"\n",
    "\n",
    "# whether to display the y-axis on a logarithmic scale\n",
    "is_log_y = False\n",
    "\n",
    "# histogram plot parameter: how to display the histogram. bar is the default\n",
    "# Allowed types are: bar, barstacked, step, stepfilled\n",
    "hist_type = \"bar\"\n",
    "\n",
    "# show the best-fit Poisson distributions on the plots\n",
    "show_fit_curve = True\n",
    "\n",
    "# image file type\n",
    "fig_ext = \".png\"\n",
    "\n",
    "# if you would like to generate graphs for only a subset of your sams, specify the subset here in a list\n",
    "# format the names as they appear in the fileName column in summary_table\n",
    "# If you want to generate graphs for all sams, use subsample = samples\n",
    "subsample = samples\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2af9e366",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Generate histogram for each SAM file\n",
    "# Note: does not display figures - for that see the calls below\n",
    "\n",
    "def fit_func(k, lamb):\n",
    "    return poisson.pmf(k, lamb)    \n",
    "\n",
    "# keeping track of the poisson distribution parameters\n",
    "parameters_list = []\n",
    "\n",
    "plt.ioff()\n",
    "\n",
    "for sample in subsample:\n",
    "    \n",
    "    # read in data\n",
    "    file_name = working_dir + \"/\" + out_dir + \"/\" + sample + \"_editDistances.txt\"\n",
    "    with open(file_name, \"r\") as file:\n",
    "        lines = [int(x.strip(\"\\n\")) for x in file.readlines()]\n",
    "    \n",
    "    \n",
    "    # poisson fit  \n",
    "    # count editDistances\n",
    "    distance_counter = Counter(lines)\n",
    "\n",
    "    # extract just the counts that are greater than exclude_counts_under\n",
    "    bins = [x for x in distance_counter.keys()]\n",
    "    counts = [distance_counter[x] for x in bins]\n",
    "    sum_counts = sum(counts)\n",
    "\n",
    "    # fit poisson curve to counts\n",
    "    # need to scale things so counts sum to 1\n",
    "    counts_scaled = [y/sum_counts for y in counts]\n",
    "    parameters, cov_matrix = curve_fit(fit_func, bins, counts_scaled, p0=0)    \n",
    "\n",
    "    # get the curve that we fit and scale it so it makes sense on the axis with the histogram\n",
    "    linspace = np.linspace(0, max(lines), max(lines) + 1)\n",
    "    expected_curve = poisson.pmf(linspace, parameters) * sum_counts\n",
    "    parameters_list.extend(parameters)\n",
    "    \n",
    "    # plot\n",
    "    fig = plt.figure()\n",
    "    ax = plt.subplot(111)\n",
    "    \n",
    "    #ax.hist(lines, max(lines)+1, log=is_log_y, histtype=hist_type)\n",
    "    ax.bar(list(distance_counter.keys()), list(distance_counter.values()), width=1.0)\n",
    "    ax.set_xlabel(\"edit distance (NM)\")\n",
    "    ax.set_ylabel(\"number of mapped reads\")\n",
    "    \n",
    "    if(show_fit_curve):\n",
    "        ax.plot(linspace, expected_curve, color=\"red\")\n",
    "    \n",
    "    if(is_log_y):\n",
    "        ax.set_ylabel(\"number of mapped reads (log)\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"number of mapped reads\")\n",
    "        ax.get_yaxis().set_major_formatter(matplotlib.ticker.FuncFormatter(lambda x, p: format(int(x), ',')))\n",
    "    \n",
    "    plt.title(sample + \" alignment edit distances\")\n",
    "    \n",
    "    fig.savefig(working_dir + \"/\" + fig_dir + \"/\" + fig_prefix + sample + fig_ext)\n",
    "    plt.close()\n",
    "\n",
    "tmp = plt.ion()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7076bea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For Poisson curves fit to the edit distance distributions, show the distribution of the lambda parameter\n",
    "\n",
    "fig, ax = plt.subplots(2, 1)\n",
    "\n",
    "ax[0].boxplot(parameters_list, vert=False)\n",
    "ax[1].scatter(parameters_list, [0.95 + random.random()/10.0 for x in parameters_list])\n",
    "ax[1].set_ylim([0.9, 1.1])\n",
    "\n",
    "ax[1].set_xlabel(\"lambda\")\n",
    "\n",
    "ax[0].yaxis.set_major_locator(ticker.NullLocator())\n",
    "ax[1].yaxis.set_major_locator(ticker.NullLocator())\n",
    "\n",
    "ax[0].set_title(\"Distribution of lamba parameters of best-fit Poisson distributions\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "421dd7be",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Display figures side-by-side\n",
    "\n",
    "fig_names = [(working_dir + \"/\" + fig_dir + \"/\" + fig_prefix + x + fig_ext) for x in subsample]\n",
    "\n",
    "figs = [plt.imread(f) for f in fig_names]\n",
    "\n",
    "ipyplot.plot_images(figs, img_width=400)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f7beb4e",
   "metadata": {},
   "source": [
    "# Comparative Histograms\n",
    "\n",
    "Sometimes, you may wish to overlay two histograms in order to compare them directly. For example, you might have aligned your fastqs to two different reference sequences. Use the cells below to display pairs of overlaid histograms.\n",
    "\n",
    "Make sure you have run the \"Run Plugin\" cells on all of the sam files you wish to use. If you are comparing alignments from the same fastq file across multiple references/settings, you should use a separate directory for `out_dir` for each, to prevent files from being overwritten.\n",
    "\n",
    "Next, create a config file with the pairs you would like to graph. This file should be a plain-text table with the following columns (separated by tabs, and including the header):\n",
    "* **fig_title**: the title of the figure\n",
    "* **fig_name**: the file name for the generated figure (not including extension)\n",
    "* **file_1**: the path to the first metrics file. It should end with _editDistances.txt\n",
    "* **label_1**: the label to use in the legend for the first file\n",
    "* **file_2**: same as **file_1**, but for the second file\n",
    "* **label_2**: same as **label_1**, but for the second file\n",
    "\n",
    "| fig_title | fig_name | file_1 | label_1 | file_2 | label_2 |\n",
    "| :-------- | :------- | :----- | :------ | :----- | :------ |\n",
    "| Example figure title | example_fig_name | /path/to/file_1_editDistances.txt | method1 | /path/to/file_2_editDistances.txt | method2 |\n",
    "\n",
    "\n",
    "Set the parameters in the \"Edit Me\" cell below and run the following cells to see your figures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23798e4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########\n",
    "# EDIT ME #\n",
    "###########\n",
    "\n",
    "# Path to the config file (see paragraph above for details)\n",
    "config_file = working_dir + \"/sam_config.txt\"\n",
    "\n",
    "# figure prefix - this is appended to the start of the figure file name when it is saved\n",
    "# use this if you want to generate multiple versions of the same figure and keep them in the same folder\n",
    "fig_prefix = \"\"\n",
    "\n",
    "# whether or not to display the y-axis on a logarithmic scale\n",
    "is_log_y = False\n",
    "\n",
    "# Note: for overlaid plots, I recommend one of the following settings\n",
    "\n",
    "# hist_type=\"step\", hist_alpha=1 (Only the outlines of the histogram will show)\n",
    "# hist_type=\"stepfilled\", hist_alpha=0.4 (Histograms will be transparent)\n",
    "\n",
    "# histogram plot parameter: how to display the histogram. bar is the default\n",
    "# Allowed types are: bar, barstacked, step, stepfilled\n",
    "hist_type = \"step\"\n",
    "\n",
    "# fill opacity\n",
    "hist_alpha = 1.0\n",
    "\n",
    "# image file type\n",
    "fig_ext = \".svg\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9158401",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Load config file\n",
    "config = pandas.read_csv(config_file, sep=\"\\t\")\n",
    "config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d78cb97",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.ioff()\n",
    "\n",
    "for index, row in config.iterrows():\n",
    "    fig=plt.figure()\n",
    "    ax=plt.subplot(111)\n",
    "    \n",
    "    with open(row[\"file_1\"], \"r\") as file:\n",
    "            lines_1 = [int(x.strip(\"\\n\")) for x in file.readlines()]\n",
    "    with open(row[\"file_2\"], \"r\") as file:\n",
    "            lines_2 = [int(x.strip(\"\\n\")) for x in file.readlines()]\n",
    "    \n",
    "    ax.hist(lines_1, max(lines_1)+1, histtype=hist_type, alpha=hist_alpha, log=is_log_y, label=row[\"label_1\"])\n",
    "    ax.hist(lines_2, max(lines_2)+1, histtype=hist_type, alpha=hist_alpha, log=is_log_y, label=row[\"label_2\"])\n",
    "    \n",
    "    ax.set_xlabel(\"edit distance (NM)\")\n",
    "    \n",
    "    if(is_log_y):\n",
    "        ax.set_ylabel(\"number of mapped reads (log)\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"number of mapped reads\")\n",
    "        ax.get_yaxis().set_major_formatter(matplotlib.ticker.FuncFormatter(lambda x, p: format(int(x), ',')))\n",
    "    \n",
    "    plt.title(row[\"fig_title\"])\n",
    "    plt.legend()\n",
    "    \n",
    "    fig.savefig(working_dir + \"/\" + fig_dir + \"/\" + fig_prefix + row[\"fig_name\"] + fig_ext)\n",
    "    plt.close()\n",
    "\n",
    "tmp = plt.ion()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1004f1f7",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# Display figures side-by-side\n",
    "\n",
    "fig_names = [(working_dir + \"/\" + fig_dir + \"/\" + fig_prefix + x + fig_ext) for x in config[\"fig_name\"]]\n",
    "\n",
    "figs = [plt.imread(f) for f in fig_names]\n",
    "\n",
    "ipyplot.plot_images(figs, img_width=400)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8efe987d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
