{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "feec0423",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Imports and other setup\n",
    "import subprocess\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas\n",
    "import os\n",
    "import linecache\n",
    "import matplotlib.ticker as mtick"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86176561",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########\n",
    "# EDIT ME #\n",
    "###########\n",
    "\n",
    "#Change paths here based on your file system\n",
    "\n",
    "#Main working directory\n",
    "working_dir = \"/Users/ahb232/Desktop/maize_stats/\"\n",
    "\n",
    "#Directory containing coverage and identity wiggle files for each chromosome\n",
    "#Files should be names coverage_X.wig or identity_X.wig\n",
    "#Where X is the name of each chromosome\n",
    "wiggle_dir = working_dir + \"/wiggles/\"\n",
    "\n",
    "#Directory containing reference fasta and gff\n",
    "reference_dir = working_dir + \"/ref/\"\n",
    "\n",
    "#Species name, for displaying on plots\n",
    "species_name = \"maize\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b83db2d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "###########\n",
    "# EDIT ME #\n",
    "###########\n",
    "\n",
    "#A list of chromosome names that we can use later\n",
    "\n",
    "# NOTE: The cell below attempts to automatically identify and sort chromosomes based on the names of the wiggle files.\n",
    "# It will sort numerical chromosome names first, and then append any other chromosomes alphabetically.\n",
    "# If you want to order the chromosomes in a different way, or use some subset of chromosomes,\n",
    "# Please edit the list below, and skip the next code cell.\n",
    "\n",
    "chrom = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8', 'chr9', 'chr10']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11ab1e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Infer chromosome names from coverage wiggle files\n",
    "for file_name in os.listdir(wiggle_dir):\n",
    "    if (file_name.startswith(\"coverage\") and file_name.endswith(\".wig\")):\n",
    "        chrom.append(file_name.replace(\"coverage_\", \"\").replace(\".wig\", \"\"))\n",
    "\n",
    "# Split names into numeric and non-numeric\n",
    "numeric_chrom = [int(x) for x in chrom if x.isnumeric()]\n",
    "alphabetical_chrom = [x for x in chrom if not x.isnumeric() ]\n",
    "\n",
    "# Sort numeric from lowest to highest, non-numeric alphabetically\n",
    "numeric_chrom.sort()\n",
    "alphabetical_chrom.sort()\n",
    "\n",
    "# Concatenate both lists to return the full list of chromosomes in the desired order\n",
    "chrom = [str(x) for x in numeric_chrom] + alphabetical_chrom\n",
    "\n",
    "print(chrom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84186d74",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get distribution of coverage for each chromosome\n",
    "\n",
    "# for constructing the dataframe\n",
    "coverage_df = \"\"\n",
    "is_first = True\n",
    "\n",
    "for c in chrom:\n",
    "    \n",
    "    # Get filepath for coverage file\n",
    "    wiggle_file = wiggle_dir + \"/coverage_\" + c + \".wig\"\n",
    "        \n",
    "    #Count the number of lines with each level of coverage\n",
    "    p1=subprocess.Popen([\"sort\", wiggle_file], stdout=subprocess.PIPE)\n",
    "    p2=subprocess.run([\"uniq\", '-c'], stdin=p1.stdout, text=True, stdout=subprocess.PIPE)\n",
    "        \n",
    "    #Reformat and remove lines that are non-numeric\n",
    "    cov = [x.split(\" \") for x in p2.stdout.split(\"\\n\")]\n",
    "    cov = [[int(x[0]), int(x[1])] for x in cov if x[0].isnumeric()]\n",
    "        \n",
    "    chr_len = sum([x[0] for x in cov])\n",
    "        \n",
    "    if is_first:\n",
    "        #Create dataframe\n",
    "        coverage_df = pandas.DataFrame({\"coverage\":[x[1] for x in cov], c:[x[0]/chr_len for x in cov]})\n",
    "        is_first=False\n",
    "    else:\n",
    "        #Add new column to existing dataframe\n",
    "        coverage_df[c] = [x[0]/chr_len for x in cov]\n",
    "\n",
    "# sort rows from lowest coverage to highest\n",
    "coverage_df = coverage_df.sort_values(by=[\"coverage\"])\n",
    "\n",
    "# get range for coverage\n",
    "min_cov = min(coverage_df[\"coverage\"])\n",
    "max_cov = max(coverage_df[\"coverage\"]+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c378faba",
   "metadata": {},
   "outputs": [],
   "source": [
    "############\n",
    "# OPTIONAL #\n",
    "############\n",
    "\n",
    "#Print coverage dataframe\n",
    "print(coverage_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb374a38",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot coverage distribution across all chromosomes\n",
    "for c in chrom:\n",
    "    plt.plot(coverage_df[\"coverage\"], coverage_df[c], label=c)\n",
    "\n",
    "plt.xticks(np.arange(min_cov, max_cov))        \n",
    "plt.legend(title=\"chromosome\", loc=\"upper left\")\n",
    "plt.xlabel(\"Coverage\")\n",
    "plt.ylabel(\"Percentage of total bases with the given coverage\")\n",
    "plt.title(\"Alignment coverage of \" + species_name + \" assemblies against reference\")\n",
    "plt.gca().yaxis.set_major_formatter(mtick.PercentFormatter(xmax=1, decimals=None, symbol='%', is_latex=False))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9082fd79",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Same information as the above graph, but cumulative\n",
    "coverage_df_cumulative = coverage_df.copy()\n",
    "\n",
    "for c in chrom:\n",
    "    coverage_df_cumulative[c] = [sum(coverage_df[c].iloc[x:len(coverage_df[c]) + 1]) for x in range(len(coverage_df[c]))]\n",
    "\n",
    "# In this graph, descending coverage is easier to read\n",
    "temp = coverage_df_cumulative.sort_values(by=[\"coverage\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eadf8fb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot cumulative coverage distribution across all chromosomes\n",
    "# Note the reversed x-axis!\n",
    "for c in chrom:\n",
    "    plt.plot(coverage_df_cumulative[\"coverage\"], coverage_df_cumulative[c], label=c)\n",
    "\n",
    "plt.xticks(np.arange(min_cov, max_cov))        \n",
    "plt.legend(title=\"chromosome\", loc=\"lower right\")\n",
    "plt.axis([max_cov-1, min_cov, 0, 1])\n",
    "plt.xlabel(\"Coverage\")\n",
    "plt.ylabel(\"Percentage of total bases with at least the given coverage\")\n",
    "plt.title(\"Cumulative  alignment coverage of \" + species_name + \" assemblies against reference\")\n",
    "plt.gca().yaxis.set_major_formatter(mtick.PercentFormatter(xmax=1, decimals=None, symbol='%', is_latex=False))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b64243b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Get distribution of identity for each chromosome\n",
    "\n",
    "# for constructing the dataframe\n",
    "identity_df = \"\"\n",
    "is_first = True\n",
    "\n",
    "for c in chrom:\n",
    "    \n",
    "    # Get filepath for identity file\n",
    "    wiggle_file = wiggle_dir + \"/identity_\" + c + \".wig\"\n",
    "        \n",
    "    #Count the number of lines with each level of identity\n",
    "    p1=subprocess.Popen([\"sort\", wiggle_file], stdout=subprocess.PIPE)\n",
    "    p2=subprocess.run([\"uniq\", '-c'], stdin=p1.stdout, text=True, stdout=subprocess.PIPE)\n",
    "        \n",
    "    #Reformat and remove lines that are non-numeric\n",
    "    idy = [x.split(\" \") for x in p2.stdout.split(\"\\n\")]\n",
    "    idy = [[int(x[0]), int(x[1])] for x in idy if x[0].isnumeric()]\n",
    "        \n",
    "    chr_len = sum([x[0] for x in idy])\n",
    "        \n",
    "    if is_first:\n",
    "        #Create dataframe\n",
    "        identity_df = pandas.DataFrame({\"identity\":[x[1] for x in idy], c:[x[0]/chr_len for x in idy]})\n",
    "        is_first=False\n",
    "    else:\n",
    "        #Add new column to existing dataframe\n",
    "        identity_df[c] = [x[0]/chr_len for x in idy]\n",
    "\n",
    "# sort rows from lowest identity to highest\n",
    "identity_df = identity_df.sort_values(by=[\"identity\"])\n",
    "\n",
    "# get range for identity\n",
    "min_idy = min(identity_df[\"identity\"])\n",
    "max_idy = max(identity_df[\"identity\"]+1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9bbc9f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "############\n",
    "# OPTIONAL #\n",
    "############\n",
    "\n",
    "#Print identity dataframe\n",
    "print(identity_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f947ed7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot identity distribution across all chromosomes\n",
    "for c in chrom:\n",
    "    plt.plot(identity_df[\"identity\"], identity_df[c], label=c)\n",
    "\n",
    "plt.xticks(np.arange(min_idy, max_idy))        \n",
    "plt.legend(title=\"chromosome\", loc=\"upper left\")\n",
    "plt.xlabel(\"identity\")\n",
    "plt.ylabel(\"Percentage of total bases with the given identity\")\n",
    "plt.title(\"Alignment identity of \" + species_name + \" assemblies against reference\")\n",
    "plt.gca().yaxis.set_major_formatter(mtick.PercentFormatter(xmax=1, decimals=None, symbol='%', is_latex=False))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18adfb80",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Same information as the above graph, but cumulative\n",
    "identity_df_cumulative = identity_df.copy()\n",
    "\n",
    "for c in chrom:\n",
    "    identity_df_cumulative[c] = [sum(identity_df[c].iloc[x:len(identity_df[c]) + 1]) for x in range(len(identity_df[c]))]\n",
    "\n",
    "# In this graph, descending identity is easier to read\n",
    "temp = identity_df_cumulative.sort_values(by=[\"identity\"], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee6ec612",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot cumulative identity distribution across all chromosomes\n",
    "# Note the reversed x-axis!\n",
    "for c in chrom:\n",
    "    plt.plot(identity_df_cumulative[\"identity\"], identity_df_cumulative[c], label=c)\n",
    "\n",
    "plt.xticks(np.arange(min_idy, max_idy))        \n",
    "plt.legend(title=\"chromosome\", loc=\"lower right\")\n",
    "plt.axis([max_idy-1, min_idy, 0, 1])\n",
    "plt.xlabel(\"identity\")\n",
    "plt.ylabel(\"Percentage of total bases with at least the given identity\")\n",
    "plt.title(\"Cumulative alignment identity of \" + species_name + \" assemblies against reference\")\n",
    "plt.gca().yaxis.set_major_formatter(mtick.PercentFormatter(xmax=1, decimals=None, symbol='%', is_latex=False))\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d98dc6bc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
